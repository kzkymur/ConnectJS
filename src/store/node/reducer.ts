import { Reducer } from 'redux';
import Action, { ActionTypes } from './actionTypes';
import { ConnectionType } from './types';
import NodeType from './nodeTypes';
import ReverseActionBranch, { OperationTypes } from './reverseActionBranch';
import Engine from './engine';

export interface State {
  nodes: NodeType[];
  nodeLatestId: number;
  connectionLatestId: number;
  connections: ConnectionType[];
  reverseActionBranch: ReverseActionBranch;
  curving: number;
  engines: Engine[];
}
export const initialState: State = {
  nodes: [],
  nodeLatestId: 0,
  connectionLatestId: 0,
  connections: [],
  reverseActionBranch: new ReverseActionBranch(),
  curving: 0.5,
  engines: [],
};

const reducer: Reducer<State, Action> = (state = initialState, action) => {
  console.log(action);
  switch (action.type) {
    case ActionTypes.add: {
      let latestId = state.nodeLatestId;
      let node = { ...action.payload.node };
      if (node.id === -1) {
        latestId++;
        node.id = latestId;
        node.name = 'node' + String(latestId);
      }
      state = {
        ...state,
        nodeLatestId: latestId,
        nodes: [ ...state.nodes, node ],
      };
      break;
    }
    case ActionTypes.delete: {
      const deletednode = state.nodes.filter(n => n.id === action.payload.id)[0];
      if (deletednode === undefined) return state;
      state = {
        ...state,
        nodes: state.nodes.filter(n => n.id !== action.payload.id),
      };
      break;
    }

    case ActionTypes.update: {
      const oldnode = state.nodes.filter(n => n.id === action.payload.node.id)[0];
      if (oldnode === undefined) return state;
      state = {
        ...state,
        nodes: state.nodes.map(c => c.id === action.payload.node.id ? action.payload.node : c),
      };
      break;
    }
    case ActionTypes.updateName: {
      const oldnode = state.nodes.filter(n => n.id === action.payload.id)[0];
      if (oldnode === undefined) return state;
      state = {
        ...state,
        nodes: state.nodes.map(c => c.id === action.payload.id ? {
          ...c,
          name: action.payload.name,
        } : c),
      };
      break;
    }
    case ActionTypes.updateSize: {
      const oldnode = state.nodes.filter(n => n.id === action.payload.id)[0];
      if (oldnode === undefined) return state;
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.id ? {
          ...n,
          width: action.payload.width,
          height: action.payload.height,
        } : n),
      };
      break;
    }
    case ActionTypes.updatePos: {
      const oldnode = state.nodes.filter(n => n.id === action.payload.id)[0];
      if (oldnode === undefined) return state;
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.id ? {
          ...n,
          top: action.payload.top,
          left: action.payload.left,
        } : n),
      };
      break;
    }

    case ActionTypes.addSocket: {
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.nodeId ? {
          ...n,
          inputsLatestId: action.payload.isInput ? n.inputsLatestId + 1 : n.inputsLatestId,
          inputs: action.payload.isInput ? [...n.inputs, {
            id: n.inputsLatestId + 1,
            type: action.payload.type,
            name: action.payload.type,
            counterId: -1,
          }] : n.inputs,
          outputsLatestId: !action.payload.isInput ? n.outputsLatestId + 1 : n.outputsLatestId,
          outputs: !action.payload.isInput ? [...n.outputs, {
            id: n.outputsLatestId + 1,
            type: action.payload.type,
            name: action.payload.type,
            counterId: -1,
          }] : n.outputs,
        } : n),
      };
      break;
    }
    case ActionTypes.deleteSocket: {
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.nodeId ? {
          ...n,
          inputs: action.payload.isInput ? n.inputs.filter(i => i.id !== action.payload.id) : n.inputs,
          outputs: !action.payload.isInput ? n.outputs.filter(o => o.id !== action.payload.id) : n.outputs,
        } : n),
      };
      break;
    }

    case ActionTypes.addConnection: {
      let latestId = state.connectionLatestId;
      let connection = { ...action.payload.connection, }
      if (connection.id === -1) {
        latestId++;
        connection.id = latestId;
      }
      state = {
        ...state,
        connections: [ ...state.connections, connection ],
        connectionLatestId: latestId,
      }
      break;
    }
    case ActionTypes.deleteConnection: {
      const deletedConnection = state.connections.filter(c => c.id === action.payload.id)[0];
      if (deletedConnection === undefined) return state;
      state = {
        ...state,
        connections: state.connections.filter(c => c.id !== action.payload.id),
      };
      break;
    }
    case ActionTypes.updateConnectionPos: {
      const oldCon = state.connections.filter(c => c.id === action.payload.id)[0];
      if (oldCon === undefined) return state;
      state = {
        ...state,
        connections: state.connections.map(c => c.id === action.payload.id ? {
          ...c,
          s: action.payload.s,
          e: action.payload.e,
        } : c),
      };
      break;
    }
    case ActionTypes.updateConnectionType: {
      const oldCon = state.connections.filter(c => c.id === action.payload.id)[0];
      if (oldCon === undefined) return state;
      state = {
        ...state,
        connections: state.connections.map(c => c.id === action.payload.id ? {
          ...c,
          type: action.payload.type,
        } : c),
      };
      break;
    }

    case ActionTypes.branch: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.branch) };
      break;
    }
    case ActionTypes.forward: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.forward) };
      break;
    }
    case ActionTypes.backward: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.backward) };
      break;
    }
    case ActionTypes.store: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.store, action.payload) };
      break;
    }

    default: return state; // 再描画しない〜〜
  }

  return { ...state, }
}
export default reducer;
