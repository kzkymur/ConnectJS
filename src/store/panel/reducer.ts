import { Reducer } from 'redux';
import Action, { ActionTypes } from './actionTypes';

interface State {
  cpIdsList: number[][];
}
export const initialState: State = {
  cpIdsList: [[]],
};

type ReducerType = Reducer<State, Action>;
const reducer: ReducerType = (state = initialState, action) => {
  return reducerLogic(state, action);
}
export default reducer;

type ReducerLogic = (state: State, action: Action) => State;
const reducerLogic: ReducerLogic = (state, action) => {
  switch (action.type) {
    case ActionTypes.openCP: {
      if(state.cpIdsList[0].indexOf(action.payload.id)>-1) {
        return state;
      } else {
        return {
          ...state,
          cpIdsList: [state.cpIdsList[0].concat([action.payload.id])],
        }
      }
    }
    case ActionTypes.closeCP: {
      return {
        ...state,
        cpIdsList: [state.cpIdsList[0].filter((_,i) => i !== action.payload.index)],
      }
    }
    case ActionTypes.closeAllCP: {
      return {
        ...state,
        cpIdsList: [[]],
      }
    }

    case ActionTypes.closeCPById: {
      return {
        ...state,
        cpIdsList: [state.cpIdsList[0].filter(id => id !== action.payload.id)],
      }
    }
    default: return state;
  }
}
