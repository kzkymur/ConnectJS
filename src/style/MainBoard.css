.mainBoard{
  position: relative;
  display: block;
  overflow: hidden;
  height: 100vh;
  width: 100vw;
  z-index: 1;
}

.connection-panel{
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: -99999999999999999;
}
