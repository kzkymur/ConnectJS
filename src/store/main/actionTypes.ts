import { Action } from 'redux';
import { DataType } from './node';
import { ConnectionType, } from './types';
import Vector from '@/utils/vector';
import { EngineType } from './engine';
import ContentType from '@/content/types';

export const ActionTypes = {
  add: "ADD",
  delete: "DELETE",
  update: "UPDATE",
  updateName: "UPDATENAME",
  updateSize: "UPDATESIZE",
  updatePos: "UPDATEPOS",
  rerender: "RERENDER",
  addSocket: "ADDSOCKET",
  deleteSocket: "DELETESOCKET",

  undo: "UNDO",
  redo: "REDO",
  mult: "MULT",

  branch: "BRANCH",
  forward: "FORWARD",
  backward: "BACKWARD",
  store: "STORE",

  addConnection: "ADDCONNECTION",
  deleteConnection: "DELETECONNECTION",
  updateConnection: "UPDATECONNECTION",
  updateConnectionPos: "UPDATECONNECTIONPOS",
  updateConnectionType: "UPDATECONNECTIONTYPE",

  addEngine: "ADDENGINE",
  deleteEngine: "DELETEENGINE",
} as const;

// Actionの型定義
interface Add extends Action {
  type: typeof ActionTypes.add;
  payload: { node: ContentType; };
}
interface Delete extends Action {
  type: typeof ActionTypes.delete;
  payload: { id: number; };
}
interface Update extends Action {
  type: typeof ActionTypes.update;
  payload: { node: ContentType; };
}
interface UpdateName extends Action {
  type: typeof ActionTypes.updateName;
  payload: {
    id: number;
    name: string;
  };
}
interface UpdateSize extends Action {
  type: typeof ActionTypes.updateSize;
  payload: {
    id: number;
    width: string;
    height: string;
  };
}
interface UpdatePos extends Action {
  type: typeof ActionTypes.updatePos;
  payload: { 
    id: number;
    top: string;
    left: string;
  };
}
interface Rerender extends Action {
  type: typeof ActionTypes.rerender;
  payload: { id: number; };
}
interface AddSocket extends Action {
  type: typeof ActionTypes.addSocket;
  payload: {
    nodeId: number;
    isInput: boolean;
    type: DataType;
  };
}
interface DeleteSocket extends Action {
  type: typeof ActionTypes.deleteSocket;
  payload: {
    nodeId: number;
    isInput: boolean;
    id: number;
  };
}

interface Undo extends Action { type: typeof ActionTypes.undo; }
interface Redo extends Action { type: typeof ActionTypes.redo; }
interface Mult extends Action {
  type: typeof ActionTypes.mult;
  payload: Action[];
}

interface Branch extends Action { type: typeof ActionTypes.branch; }
interface Forward extends Action { type: typeof ActionTypes.forward; }
interface Backward extends Action { type: typeof ActionTypes.backward; }
interface Store extends Action {
  type: typeof ActionTypes.store;
  payload: Action[];
}

interface AddConnection extends Action { 
  type: typeof ActionTypes.addConnection; 
  payload: {
    connection: ConnectionType;
  };
}
interface DeleteConnection extends Action { 
  type: typeof ActionTypes.deleteConnection; 
  payload: { id: number; };
}
interface UpdateConnection extends Action { 
  type: typeof ActionTypes.updateConnection; 
  payload: {
    connection: ConnectionType;
  };
}
interface UpdateConnectionPos extends Action { 
  type: typeof ActionTypes.updateConnectionPos;
  payload: {
    id: number;
    s: Vector;
    e: Vector;
  };
}
interface UpdateConnectionType extends Action { 
  type: typeof ActionTypes.updateConnectionType;
  payload: {
    id: number;
    type: DataType;
  };
}
interface AddEngine extends Action { 
  type: typeof ActionTypes.addEngine; 
  payload: { engine: EngineType; };
}
interface DeleteEngine extends Action { 
  type: typeof ActionTypes.deleteEngine; 
  payload: { id: number; };
}

type NodeAction =
  Add | Delete |
  Update | UpdateName | UpdateSize | UpdatePos | Rerender |
  AddSocket | DeleteSocket |
  Undo | Redo | Mult |
  Branch | Forward | Backward | Store |
  AddConnection | DeleteConnection |
  UpdateConnection | UpdateConnectionPos | UpdateConnectionType |
  AddEngine | DeleteEngine;
export default NodeAction;
