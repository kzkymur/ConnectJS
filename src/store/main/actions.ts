import Action, { ActionTypes } from './actionTypes';
import { DataType } from './node';
import { ConnectionType, } from './types';
import Vector from '@/utils/vector';
import { EngineType } from './engine'
import ContentType from '@/content/types';

export const addAction = (node: ContentType): Action => ({
  type: ActionTypes.add,
  payload: { node, },
});
export const deleteAction = (nodeId: number): Action => ({
  type: ActionTypes.delete,
  payload: { id: nodeId, },
});

export const updateAction = (node: ContentType): Action => ({
  type: ActionTypes.update,
  payload: { node, },
});
export const updateNameAction = (id: number, name: string): Action => ({
  type: ActionTypes.updateName,
  payload: { id, name, },
});
export const updateSizeAction = (id: number, width: string, height: string): Action => ({
  type: ActionTypes.updateSize,
  payload: { id, width, height, },
});
export const updatePosAction = (id: number, top: string, left: string): Action => ({
  type: ActionTypes.updatePos,
  payload: { id, top, left, },
});
export const rerenderAction = (id: number): Action => ({
  type: ActionTypes.rerender,
  payload: { id, },
});
export const addSocketAction = (nodeId: number, isInput: boolean, type: DataType): Action => ({
  type: ActionTypes.addSocket,
  payload: { nodeId, isInput, type },
});
export const deleteSocketAction = (nodeId: number, isInput: boolean, id: number): Action => ({
  type: ActionTypes.deleteSocket,
  payload: { nodeId, isInput, id },
});
export const undoAction = (): Action => ({ type: ActionTypes.undo, });
export const redoAction = (): Action => ({ type: ActionTypes.redo, });
export const multAction = (actions: Action[]): Action => ({
  type: ActionTypes.mult,
  payload: actions,
});

export const branchAction = (): Action => ({ type: ActionTypes.branch, });
export const forwardAction = (): Action => ({ type: ActionTypes.forward, });
export const backwardAction = (): Action => ({ type: ActionTypes.backward, });
export const storeAction = (actions: Action[]): Action => ({
  type: ActionTypes.store,
  payload: actions,
});

export const addConnectionAction = (connection: ConnectionType): Action => ({
  type: ActionTypes.addConnection,
  payload: { connection, },
});
export const updateConnectionAction = (connection: ConnectionType): Action => ({ 
  type: ActionTypes.updateConnection, 
  payload: { connection, },
});
export const updateConnectionTypeAction = (id: number, type: DataType): Action => ({ 
  type: ActionTypes.updateConnectionType,
  payload: { id, type },
});
export const updateConnectionPosAction = (id: number, s: Vector, e: Vector): Action => ({ 
  type: ActionTypes.updateConnectionPos,
  payload: { id, s, e },
});
export const deleteConnectionAction = (id: number): Action => ({
  type: ActionTypes.deleteConnection,
  payload: { id, },
});

export const addEngine = (engine: EngineType): Action => ({
  type: ActionTypes.addEngine,
  payload: { engine, },
});
export const deleteEngine = (id: number): Action => ({
  type: ActionTypes.deleteEngine,
  payload: { id, },
});
