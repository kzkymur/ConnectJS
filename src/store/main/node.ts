import { initBaseWidth, initBaseHeight } from '@/config';
import ContentType from '@/content/types';
import { ConnectionType } from './types';
import Action from './actionTypes';
import { rerenderAction } from './actions';

export class Node<To = any, Args extends { [arg: string]: unknown; } = any> {
  id: number;
  func: (args: Args) => void;
  name: string;
  outputs: Socket[];
  inputs: Socket[];
  output?: To;
  inputsLatestId: number;
  outputsLatestId: number;
  args: { [Key in keyof Args]?: Args[Key]; };
  readonly keys: Array<keyof Args>;
  updateStore?: () => void;
  tos: Record<number, {
    node: ContentType;
    key: string;
  }>;
  constructor (dummyFunc: (args: Args) => To, keys: Array<keyof Args>) {
    this.id = -1;
    this.name = '';
    this.outputs = [];
    this.inputs = [];
    this.outputsLatestId = 0;
    this.inputsLatestId = 0;
    this.keys = keys;
    this.args = {};
    this.func = dummyFunc;
    this.tos = [];
  }
  updateName (name: string) {
    this.name = name;
    return this;
  }
  set function (func: (args: Args) => To) {
    this.func = function (args: Args) {
      const output = func(args);
      if (this.output === output) return;
      this.output = output;
      this.updateStore!();
      this.emit();
    };
  }
  emit (...ids: number[]) {
    let targets;
    if (ids.length === 0) targets = this.tos;
    else {
      targets = {};
      ids.forEach(id => targets[id] = {});
    };
    for (const i in targets) {
      const obj = {};
      Object.defineProperty(obj, this.tos[i].key, { value: this.output, enumerable: true, });
      this.tos[i].node.arg = obj;
    };
  }
  set arg (args: { [key in keyof Args]?: Args[key] }) {
    for (const key in args) {
      this.args[key] = args[key];
    }
    if (this.isFullArgs(this.args)) this.func!(this.args);
  }
  isFullArgs (args: { [Key in keyof Args]?: Args[Key] }): args is Args {
    for (const key of this.keys) if (args[key] === undefined) return false;
    return true;
  }
  addTos (node: ContentType, key: string) {
    if (this.tos[node.id] !== undefined) throw new Error('This node id has already been registered');
    this.tos[node.id] = { key, node };
    this.emit(node.id);
    return this;
  }
  updateTos (node: ContentType) {
    if (this.tos[node.id] === undefined) throw new Error('This node id has been unused');
    this.tos[node.id] = { key: this.tos[node.id].key, node };
    return this;
  }
  delTos (id: number) {
    if (this.tos[id] === undefined) throw new Error('This node id has been unused');
    delete this.tos[id];
    return this;
  }
  set dispatch (dispatch: (action: Action) => void) {
    this.updateStore = () => dispatch(rerenderAction(this.id));
  }
  rerender () { return this; }
}

export class MovableNode<To = any, Args extends { [arg: string]: unknown; } = { [arg: string]: unknown; }> extends Node<To, Args> {
  top: string;
  left: string;
  constructor (dummyFunc: (args: Args) => To, keys: Array<keyof Args>) {
    super(dummyFunc, keys);
    this.top = `${String(40 + 20 * Math.random())}%`;
    this.left = `${String(40 + 20 * Math.random())}%`;
  }
  updatePos (top: string, left: string) {
    this.top = top;
    this.left = left;
    return this;
  }
}
export function isMovable (arg: Node): arg is MovableNode {
  return arg instanceof MovableNode;
}

export class ResizableNode<To = any, Args extends { [arg: string]: unknown; } = { [arg: string]: unknown; }> extends MovableNode<To, Args> {
  width: string;
  height: string;
  constructor (dummyFunc: (args: Args) => To, keys: Array<keyof Args>) {
    super(dummyFunc, keys);
    this.width = `${initBaseWidth}px`;
    this.height = `${initBaseHeight}px`;
  }
  updateSize (width: string, height: string) {
    this.width = width;
    this.height = height;
    return this;
  }
}
export function isResizable (arg: Node): arg is ResizableNode {
  return arg instanceof ResizableNode;
}

type Data = {
  Number: number;
}
export type DataType = Data[keyof Data];

export interface Socket {
  type: DataType;
  id: number;
  name: string;
  connection?: ConnectionType;
}
