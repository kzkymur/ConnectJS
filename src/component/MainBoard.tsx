import React, { useState, useEffect } from 'react'; 
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '@/store';
import { addConnectionAction } from '@/store/node/actions';
import { openCPAction } from '@/store/panel/actions';
import { Content, Connection as ConnectionType } from '@/store/node/types';
import Base from './Base';
import ControllPanel from './ControllPanel';
import Connection from './Connection';
import { getIndex } from '@/utils';
import style from '@/style/MainBoard.css';

type BaseIdRefType = {
  id: number;
  ref: React.RefObject<HTMLDivElement>;
}

const MainBoard: React.FC = () => {
  const props = useSelector((state: RootState) => state.nodeReducer);
  const cpIdsList = useSelector((state: RootState) => state.panelReducer.cpIdsList);
  const [baseIdRefs, setBaseIdRefs] = useState<BaseIdRefType[]>([]);
  const [cpIndexes, setCPIndexes] = useState<number[]>([]);

  const dispatch = useDispatch();
  const openCPFunc = (id: number) => dispatch(openCPAction(id));
  const addConnection = (c: ConnectionType) => dispatch(addConnectionAction(c));

  // refのsetup
  const addBIR = (id: number) => [...baseIdRefs, {id: id, ref: React.createRef<HTMLDivElement>()}];
  const removeBIR = (id: number) => baseIdRefs.filter(bmi=>bmi.id!==id);

  // controlPanel系
  const createOpenCP = (id: number) => {
    const openCP = () => {
      const currentIndex = cpIdsList[0].indexOf(id);
      if (currentIndex === -1) {
        openCPFunc(id);
        createSetCPIndex(0)(cpIdsList[0].length);
      } else if (currentIndex !== cpIndexes[0]){
        createSetCPIndex(0)(currentIndex);
      }
    }
    return openCP;
  }
  const createSetCPIndex = (cpIndex: number) => {
    const setCPIndex = (newIndex: number) => {setCPIndexes( cpIndexes.map( (oldIndex, i) => i === cpIndex ? newIndex : oldIndex ))}
    return setCPIndex
  }

  // 状態確認系
  useEffect(()=>{
    let newBaseIdRefs = [...baseIdRefs];
    let flag = false;
    props.contents.forEach(c=>{
      const i = getIndex(newBaseIdRefs, c.id);
      if (i === -1) { flag = true; newBaseIdRefs = addBIR(c.id); }
    });
    newBaseIdRefs.forEach(bir=>{
      const i = getIndex(props.contents, bir.id);
      if (i === -1) { flag = true; newBaseIdRefs = removeBIR(bir.id); }
    });
    if (flag) setBaseIdRefs(newBaseIdRefs);
  }, [props.contents]);
  useEffect(()=>{
    for (const i in cpIdsList) {
      if (cpIndexes[i] === undefined) {
        return setCPIndexes([...cpIndexes, 0]);
      }
    }
  }, [cpIdsList]);

  return (
    <div className={style.mainBoard}>
      {baseIdRefs.map((baseIdRef: BaseIdRefType)=>{
        const c = props.contents.filter(c=>c.id === baseIdRef.id)[0];
        if (c===undefined) return null;
        const baseProps = {
          property: c, 
          openCP: createOpenCP(c.id),
        }
        return <Base key={c.id}{...baseProps}/>
      })}
      {cpIdsList.map((ids: number[], i)=>{
        if(ids[0]===undefined) return;
        let properties: Content[] = [];
        ids.forEach(id=>{
          let c = props.contents.filter(c=>c.id===id)[0];
          properties.push(c);
        })
        const cpProps = { 
          properties: properties, 
          index: cpIndexes[i],
          setIndex: createSetCPIndex(i),
        }
        return <ControllPanel {...cpProps} key={i}/>
      })}
      <svg className={style.connectionPanel}>
      </svg>
    </div>
  )
}

export default MainBoard;
