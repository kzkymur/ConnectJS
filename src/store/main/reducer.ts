import { Reducer } from 'redux';
import Action, { ActionTypes } from './actionTypes';
import { isMovable, isResizable } from './node';
import { ConnectionType } from './types';
import ReverseActionBranch, { OperationTypes } from './reverseActionBranch';
import { EngineType } from './engine';
import ContentType from '@/content/types';

export interface State {
  nodes: ContentType[];
  nodeLatestId: number;
  connections: ConnectionType[];
  connectionLatestId: number;
  reverseActionBranch: ReverseActionBranch;
  curving: number;
  engines: EngineType[];
  engineLatestId: number;
}
export const initialState: State = {
  nodes: [],
  nodeLatestId: 0,
  connectionLatestId: 0,
  connections: [],
  reverseActionBranch: new ReverseActionBranch(),
  curving: 0.5,
  engines: [],
  engineLatestId: 0,
};

const reducer: Reducer<State, Action> = (state = initialState, action) => {
  console.log(action);
  switch (action.type) {
    case ActionTypes.add: {
      const latestId = state.nodeLatestId + 1;
      const node = action.payload.node;
      if (node.id === -1) {
        node.id = latestId;
        node.name = 'node' + String(latestId);
      }
      state = {
        ...state,
        nodeLatestId: latestId,
        nodes: [ ...state.nodes, node ],
      };
      break;
    }
    case ActionTypes.delete: {
      state = {
        ...state,
        nodes: state.nodes.filter(n => n.id !== action.payload.id),
      };
      break;
    }

    case ActionTypes.update: {
      state = {
        ...state,
        nodes: state.nodes.map(c => c.id === action.payload.node.id ? action.payload.node : c),
      };
      break;
    }
    case ActionTypes.updateName: {
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.id ? n.updateName(action.payload.name) : n),
      };
      break;
    }
    case ActionTypes.updateSize: {
      const { width, height } = action.payload;
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.id && isResizable(n) ? n.updateSize(width, height) : n),
      };
      break;
    }
    case ActionTypes.updatePos: {
      const { top, left } = action.payload;
      state = {
        ...state,
        nodes: state.nodes.map(n => n.id === action.payload.id && isMovable(n) ? n.updatePos(top, left) : n),
      };
      break;
    }

    // case ActionTypes.addSocket: {
    //   state = {
    //     ...state,
    //     nodes: state.nodes.map(n => n.id === action.payload.nodeId ? {
    //       ...n,
    //       inputsLatestId: action.payload.isInput ? n.inputsLatestId + 1 : n.inputsLatestId,
    //       inputs: action.payload.isInput ? [...n.inputs, {
    //         id: n.inputsLatestId + 1,
    //         type: action.payload.type,
    //         name: String(action.payload.type),
    //         counterId: -1,
    //       }] : n.inputs,
    //       outputsLatestId: !action.payload.isInput ? n.outputsLatestId + 1 : n.outputsLatestId,
    //       outputs: !action.payload.isInput ? [...n.outputs, {
    //         id: n.outputsLatestId + 1,
    //         type: action.payload.type,
    //         name: String(action.payload.type),
    //         counterId: -1,
    //       }] : n.outputs,
    //     } : n),
    //   };
    //   break;
    // }
    // case ActionTypes.deleteSocket: {
    //   state = {
    //     ...state,
    //     nodes: state.nodes.map(n => n.id === action.payload.nodeId ? {
    //       ...n,
    //       inputs: action.payload.isInput ? n.inputs.filter(i => i.id !== action.payload.id) : n.inputs,
    //       outputs: !action.payload.isInput ? n.outputs.filter(o => o.id !== action.payload.id) : n.outputs,
    //     } : n),
    //   };
    //   break;
    // }

    case ActionTypes.addConnection: {
      const latestId = state.connectionLatestId + 1;
      const connection = { ...action.payload.connection, }
      if (connection.id === -1) connection.id = latestId;
      const setTo = (node: ContentType) => {
        const toNode = state.nodes.find(node => node.id === connection.toNodeId);
        console.log({toNode});
        if (toNode !== undefined) return node.addTos(toNode, toNode.keys[connection.toSocketId-1]);
        return node;
      }
      state = {
        ...state,
        connections: [ ...state.connections, connection ],
        connectionLatestId: latestId,
        nodes: state.nodes.map(node => node.id === connection.fromNodeId ? setTo(node) : node)
      }
      break;
    }
    case ActionTypes.deleteConnection: {
      state = {
        ...state,
        connections: state.connections.filter(c => c.id !== action.payload.id),
      };
      break;
    }
    case ActionTypes.updateConnectionPos: {
      state = {
        ...state,
        connections: state.connections.map(c => c.id === action.payload.id ? {
          ...c,
          s: action.payload.s,
          e: action.payload.e,
        } : c),
      };
      break;
    }
    case ActionTypes.updateConnectionType: {
      state = {
        ...state,
        connections: state.connections.map(c => c.id === action.payload.id ? {
          ...c,
          type: action.payload.type,
        } : c),
      };
      break;
    }

    case ActionTypes.branch: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.branch) };
      break;
    }
    case ActionTypes.forward: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.forward) };
      break;
    }
    case ActionTypes.backward: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.backward) };
      break;
    }
    case ActionTypes.store: {
      state = { ...state, reverseActionBranch: state.reverseActionBranch.operate(OperationTypes.store, action.payload) };
      break;
    }

    case ActionTypes.addEngine: {
      const latestId = state.engineLatestId + 1;
      const engine = { ...action.payload.engine };
      if (engine.id === -1) engine.id = latestId;
      state = {
        ...state,
        engineLatestId: latestId,
        engines: [ ...state.engines, engine ],
      };
      break;
    }
    case ActionTypes.deleteEngine: {
      state = {
        ...state,
        engines: state.engines.filter(e => e.id !== action.payload.id),
      };
      break;
    }

    default: return state; // 再描画しない〜〜
  }

  return { ...state, }
}
export default reducer;
