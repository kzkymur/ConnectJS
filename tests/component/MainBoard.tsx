import { useSelector, useDispatch } from 'react-redux'
import renderer from 'react-test-renderer';
import { initialState as initNodeState } from '@/store/main/reducer';
import { initialState as initPanelState } from '@/store/panel/reducer';
import MainBoard from '@/component/MainBoard';

jest.mock('react-redux')
const useSelectorMock = useSelector as jest.Mock
const useDispatchMock = useDispatch as jest.Mock

describe('<MainBoard />', () => {
  test('render', () => {
    useSelectorMock.mockReturnValueOnce(initNodeState).mockReturnValue(initPanelState.cpIdsList)
    useDispatchMock.mockReturnValue(jest.fn())
    const component = renderer.create( <MainBoard/> );
    expect(component).toMatchSnapshot();
  });
});
