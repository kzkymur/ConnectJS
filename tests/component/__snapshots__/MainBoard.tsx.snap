// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainBoard /> render 1`] = `
<div
  className="mainBoard"
>
  <svg>
    <path
      d="M 0 0 Q 0 0 0 0 T 0 0"
    />
  </svg>
</div>
`;
